src/hashset.h

```diff
@@ -0,0 +1,188 @@
+/* Iterator. To be able to use a stack-allocated iterator, it's not opaque. The
+ * fields in this struct should not be accessed though. Just use the hashset
+ * iterator functions. */
+typedef struct {
+    hashset *hashset;
+    hashsetBucket *bucket;
+    long index;
+    int table;
+    int pos_in_bucket;
+    uint64_t safe : 1;
+    union {
+        /* Unsafe iterator fingerprint for misuse detection. */
+        uint64_t fingerprint : 63;
+        /* Safe iterator temporary storage for bucket chain compaction. */
+        uint64_t last_seen_size : 63;
+    };
+} hashsetIterator;
+
+/* Position, used by some hashset functions such as two-phase insert and delete.
+ * This is not opaque, to be able to place it on the stack, but don't access the
+ * fields. Only use it with the hashset functions. */
+typedef struct {
+    void *bucket;
+    uint16_t pos_in_bucket;
+    uint16_t table_index;
+} hashsetPosition;
+
+/* --- Prototypes --- */
+
+/* Hash function (global seed) */
+void hashsetSetHashFunctionSeed(const uint8_t *seed);
+uint8_t *hashsetGetHashFunctionSeed(void);
+uint64_t hashsetGenHashFunction(const char *buf, size_t len);
+uint64_t hashsetGenCaseHashFunction(const char *buf, size_t len);
+
+/* Global resize policy */
+void hashsetSetResizePolicy(hashsetResizePolicy policy);
+
+/* Hashset instance */
+hashset *hashsetCreate(hashsetType *type);
+void hashsetRelease(hashset *s);
+void hashsetEmpty(hashset *s, void(callback)(hashset *));
+hashsetType *hashsetGetType(hashset *s);
+void *hashsetMetadata(hashset *s);
+size_t hashsetSize(hashset *s);
+size_t hashsetBuckets(hashset *s);
+size_t hashsetChainedBuckets(hashset *s, int table);
+size_t hashsetMemUsage(hashset *s);
+void hashsetPauseAutoShrink(hashset *s);
+void hashsetResumeAutoShrink(hashset *s);
+int hashsetIsRehashing(hashset *s);
+int hashsetIsRehashingPaused(hashset *s);
+void hashsetRehashingInfo(hashset *s, size_t *from_size, size_t *to_size);
+int hashsetRehashMicroseconds(hashset *s, uint64_t us);
+int hashsetExpand(hashset *s, size_t size);
+int hashsetTryExpand(hashset *s, size_t size);
+int hashsetExpandIfNeeded(hashset *s);
+int hashsetShrinkIfNeeded(hashset *s);
+hashset *hashsetDefragTables(hashset *s, void *(*defragfn)(void *));
+
+/* Elements */
+int hashsetFind(hashset *s, const void *key, void **found);
+void **hashsetFindRef(hashset *s, const void *key);
+int hashsetAdd(hashset *s, void *element);
+int hashsetAddOrFind(hashset *s, void *element, void **existing);
+int hashsetFindPositionForInsert(hashset *s, void *key, hashsetPosition *position, void **existing);
+void hashsetInsertAtPosition(hashset *s, void *element, hashsetPosition *position);
+int hashsetReplace(hashset *s, void *element);
+int hashsetPop(hashset *s, const void *key, void **popped);
+int hashsetDelete(hashset *s, const void *key);
+void **hashsetTwoPhasePopFindRef(hashset *s, const void *key, hashsetPosition *position);
+void hashsetTwoPhasePopDelete(hashset *s, hashsetPosition *position);
```

Comment:
Can these functions be eliminated if `hashsetFindRef()` returns a `hashsetPosition`?  I think so.

Also, in case I'm missing something, and we really need this, I think it should avoid returning `void**`.  Instead, the caller can use `hashsetGetAtPosition()` as mentioned above and also have the security of the epoch check.

The `hashsetTwoPhasePopDelete` now simply becomes `hashsetDeleteAtPosition(position)`.  This also avoids the need to pause rehashing - as the epoch will be checked.  If the caller is going to go off and do other things (and the epoch might change) it's probably worth the cost of doing another lookup.

Reply:
This is used in dbGenericDelete, where we currently use dictTwoPhaseUnlinkFind and dictTwoPhaseUnlinkDelete. I think that one is equally unsafe.

I believe we need to pause rehashing, but we can let the caller do it explicitly. The only think that I believe can happen between the two phases where this is used in db.c is that an element is replaced by another one, which doesn't need to bump the epoch (if we introduce an epoch).